### PR TITLE
Lru Memory GC

### PR DESCRIPTION
**Pull request by wu-hui**
see: https://github.com/firebase/firebase-js-sdk/pull/6943 and go/firestore-memory-lru

### CODE REVIEW

File: Firestore/core/src/api/settings.h [hunk 1]
```diff
@@ -131,14 +133,78 @@ class PersistentCacheSettings : public LocalCacheSettings {
...
+                         const MemoryGargabeCollectorSettings& rhs);
+  virtual size_t Hash() const = 0;
+
+  MemoryGcKind kind() const {
+    return kind_;
+  }
+
+ protected:
+  explicit MemoryGargabeCollectorSettings(MemoryGcKind kind)
+      : kind_(std::move(kind)) {
```
Comment (ehsannas) [inline Firestore/core/src/api/settings.h:150]: IIRC enum values are unsigned integers. There's probably little/no performance improvement for move vs copy for an integer. No need to change it though.

File: Firestore/Swift/Tests/Integration/DatabaseTests.swift [hunk 2]
```diff
@@ -53,6 +53,32 @@ import FirebaseFirestore
...
+      let settings = db.settings
+      settings
+        .cacheSettings =
+        MemoryCacheSettings(garbageCollectorSettings: MemoryEagerGCSetting())
+      db.settings = settings
+
+      try await db.document("coll/doc").setData(["foo": "bar"])
+      let result = try? await db.document("coll/doc").getDocument(source: .cache)
+      XCTAssertNil(result)
+    }
```
Comment (ehsannas) [inline Firestore/Swift/Tests/Integration/DatabaseTests.swift:80]: Nice!

**Review by ehsannas — APPROVED** [review]
Approved with a few nits.

File: Firestore/Example/Tests/SpecTests/json/collection_spec_test.json [hunk 3]
```diff
@@ -46,6 +46,7 @@
         "watchEntity": {
           "docs": [
             {
+              "createTime": 0,
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/Example/Tests/SpecTests/json/collection_spec_test.json:49]: I think it passes.

File: Firestore/Source/API/FIRLocalCacheSettings.mm [hunk 4]
```diff
@@ -82,6 +84,94 @@ - (instancetype)initWithSizeBytes:(NSNumber *)size {
...
+}
+
+- (const MemoryEagerGcSettings &)internalSettings {
+  return _internalSettings;
+}
+
+- (instancetype)init {
+  self = [super init];
+  self.internalSettings = MemoryEagerGcSettings{};
+  return self;
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/Source/API/FIRLocalCacheSettings.mm:123]: Done.

File: Firestore/Source/Public/FirebaseFirestore/FIRLocalCacheSettings.h [hunk 5]
```diff
@@ -57,6 +57,63 @@ NS_SWIFT_NAME(PersistentCacheSettings)
...
+ *
+ * The two cache types supported are `MemoryEagerGCSettings` and `MemoryLRUGCSettings`. Custom
+ * implementation is not supported.
+ */
+NS_SWIFT_NAME(MemoryGarbageCollectorSettings)
+@protocol FIRMemoryGarbageCollectorSettings
+@end
+
+/**
+ * Configures the SDK to use a eager garbage collector for memory cache.
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/Source/Public/FirebaseFirestore/FIRLocalCacheSettings.h:71]: Done.

File: Firestore/Source/Public/FirebaseFirestore/FIRLocalCacheSettings.h [hunk 6]
```diff
@@ -57,6 +57,63 @@ NS_SWIFT_NAME(PersistentCacheSettings)
...
+ * 100MB.
+ *
+ * To use, create an instance using one of the initializers, then initialize
+ * `MemoryCacheSettings` with this instance.
+ */
+NS_SWIFT_NAME(MemoryLRUGCSettings)
+@interface FIRMemoryLRUGCSettings : NSObject <NSCopying, FIRMemoryGarbageCollectorSettings>
+
+/**
+ * Creates an instnace of `FIRMemoryLRUGCSettings`, with default target cache size 100MB.  SDK will
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/Source/Public/FirebaseFirestore/FIRLocalCacheSettings.h:104]: Done.

File: Firestore/core/src/api/settings.h (hunk 1 above)
Comment (wu-hui) [reply Firestore/core/src/api/settings.h:150]: Makes sense.

File: Firestore/Example/Tests/SpecTests/json/perf_spec_test.json
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/Example/Tests/SpecTests/json/perf_spec_test.json:1]: It's some tests repeating hundreds of steps, so the file is pretty big.

File: Firestore/Source/Public/FirebaseFirestore/FIRLocalCacheSettings.h [hunk 7]
```diff
@@ -57,6 +57,63 @@ NS_SWIFT_NAME(PersistentCacheSettings)
...
+ * run garbage collection if the current cache size is larger than 100MB.
+ */
+- (instancetype)init;
+
+/**
+ * Creates an instnace of `FIRMemoryLRUGCSettings`, with a custom target cache size. SDK will
```
[the remark this answers is not in the document]
Comment (wu-hui) [reply Firestore/Source/Public/FirebaseFirestore/FIRLocalCacheSettings.h:110]: Done.

**Review by wu-hui — COMMENTED** [review]
Thanks!

**Pull request merged**